File: Awful.apk/src/main/java/com/ferg/awfulapp/task/ThreadListRequest.java
package com.ferg.awfulapp.task;

import android.content.Context;
import android.net.Uri;
import android.support.annotation.Nullable;

import com.ferg.awfulapp.announcements.AnnouncementsManager;
import com.ferg.awfulapp.constants.Constants;
import com.ferg.awfulapp.messages.PmManager;
import com.ferg.awfulapp.thread.AwfulForum;
import com.ferg.awfulapp.thread.AwfulPagedItem;
import com.ferg.awfulapp.util.AwfulError;

import org.jsoup.nodes.Document;

/**
 * Created by matt on 8/7/13.
 */
public class ThreadListRequest extends AwfulStrippedRequest<Void> {

    public static final Object REQUEST_TAG = new Object();

    private int forumId, page;

    public ThreadListRequest(Context context, int forumId, int page) {
        // TODO: 19/09/2016 decide whether to handle all USERCP requests as bookmark urls (and do the PmManager calls a different way)
        // I'm so sorry, thanks for forcing me to do this JAVA
        super(context, forumId != Constants.USERCP_ID ? Constants.FUNCTION_FORUM
                : page == 1 ? Constants.FUNCTION_USERCP
                : Constants.FUNCTION_BOOKMARK);
        this.forumId = forumId;
        this.page = page;
    }


    @Override
    public Object getRequestTag() {
        return REQUEST_TAG;
    }


    @Override
    protected String generateUrl(Uri.Builder urlBuilder) {
        if (forumId != Constants.USERCP_ID) {
            urlBuilder.appendQueryParameter(Constants.PARAM_FORUM_ID, Integer.toString(forumId));
        }
        urlBuilder.appendQueryParameter(Constants.PARAM_PAGE, Integer.toString(page));
        return urlBuilder.build().toString();
    }

    @Override
    protected Void handleResponse(Document doc) throws AwfulError {
        int lastPage = AwfulPagedItem.parseLastPage(doc);
        handleStrippedResponse(doc, page, lastPage);
        return null;
    }

    @Override
    Void handleStrippedResponse(Document doc, @Nullable Integer currentPage, @Nullable Integer totalPages) throws AwfulError {
        int thisPage = (currentPage == null) ? page : currentPage;
        int lastPage = (totalPages == null) ? thisPage : totalPages;
        // TODO: legacy try/catch - work out what this is meant to be catching exactly, and if we can ditch it
        try {
            if (forumId == Constants.USERCP_ID) {
                AwfulForum.parseUCPThreads(doc, page, lastPage, getContentResolver());
                PmManager.parseUcpPage(doc);
            } else {
                AwfulForum.parseThreads(forumId, page, lastPage, doc, getContentResolver());
                AnnouncementsManager.getInstance().parseForumPage(doc);
            }
        } catch (Exception e) {
            e.printStackTrace();
            throw new AwfulError();
        }
        return null;
    }

    @Override
    protected boolean handleError(AwfulError error, Document doc) {
        return error.isCritical();
    }
}

File: Awful.apk/src/main/java/com/ferg/awfulapp/task/ThreadListRequest.kt
package com.ferg.awfulapp.task

import android.content.Context
import android.net.Uri
import com.ferg.awfulapp.announcements.AnnouncementsManager
import com.ferg.awfulapp.constants.Constants.*
import com.ferg.awfulapp.messages.PmManager
import com.ferg.awfulapp.thread.AwfulForum
import com.ferg.awfulapp.thread.AwfulPagedItem
import com.ferg.awfulapp.util.AwfulError
import org.jsoup.nodes.Document

/**
 * Fetch the threads on a certain [page] of the forum with the given [forumId].
 *
 * This loads and parses the page, updating the cache database as appropriate,
 * depending on whether a normal or bookmarks page was loaded.
 *
 * This request also hands the page off to other parsers, e.g. for announcements
 * and private messages, to scrape any updated information the page contains.
 */
class ThreadListRequest(context: Context, private val forumId: Int, private val page: Int)
    : AwfulStrippedRequest<Void?>(context, when {
        forumId != USERCP_ID -> FUNCTION_FORUM
        page == 1 -> FUNCTION_USERCP
        else -> FUNCTION_BOOKMARK
    }) {
// TODO: 19/09/2016 decide whether to handle all USERCP requests as bookmark urls (and do the PmManager calls a different way)

    override val requestTag: Any
        get() = REQUEST_TAG


    override fun generateUrl(urlBuilder: Uri.Builder?): String {
        with(urlBuilder!!) {
            if (forumId != USERCP_ID) appendQueryParameter(PARAM_FORUM_ID, forumId.toString())
            appendQueryParameter(PARAM_PAGE, page.toString())
            return build().toString()
        }
    }

    @Throws(AwfulError::class)
    override fun handleResponse(doc: Document): Void? {
        val lastPage = AwfulPagedItem.parseLastPage(doc)
        handleStrippedResponse(doc, page, lastPage)
        return null
    }

    @Throws(AwfulError::class)
    override fun handleStrippedResponse(document: Document, currentPage: Int?, totalPages: Int?): Void? {
        val thisPage = currentPage ?: page
        val lastPage = totalPages ?: thisPage
        // TODO: legacy try/catch - work out what this is meant to be catching exactly, and if we can ditch it
        try {
            // parse the threads on the page, and also check for announcements/PMs depending on where they appear
            if (forumId == USERCP_ID) {
                AwfulForum.parseUCPThreads(document, page, lastPage, contentResolver)
                PmManager.parseUcpPage(document)
            } else {
                AwfulForum.parseThreads(forumId, page, lastPage, document, contentResolver)
                AnnouncementsManager.getInstance().parseForumPage(document)
            }
        } catch (e: Exception) {
            e.printStackTrace()
            throw AwfulError()
        }
        return null
    }


    companion object {
        val REQUEST_TAG = Any()
    }
}
